pkg/utilization/core/writers.go

```diff
@@ -0,0 +1,179 @@
+	stop     bool
+	f        *os.File
+	metadata *Metadata
+	// increment specifies by how much the line count should be incremented,
+	// this is configured when batching multiple lines into a single write operation.
+	increment int
+}
+
+// NewFileWriter creates a new FileWriter that writes to a file in the specified directory with the given metadata.
+func NewFileWriter(dir string, metadata *Metadata, increment int) (*FileWriter, error) {
+	filename := fmt.Sprintf("%s_%s_%s.%s", metadata.Workflow, metadata.Step, metadata.Execution, metadata.Format)
+	f, err := initFile(dir, filename)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	return &FileWriter{
+		f:         f,
+		metadata:  metadata,
+		increment: increment,
+	}, nil
+}
+
+// initFile creates a new file in the specified directory with the given name, reserves space for metadata,
+// and moves the cursor to the start of the data section.
+func initFile(dir, name string) (*os.File, error) {
+	// Ensure the metrics directory exists, creating it if necessary.
+	if err := os.MkdirAll(dir, 0755); err != nil {
+		return nil, errors.Wrapf(err, "failed to create %q directory", dir)
+	}
+
+	// Create the file inside the metrics directory.
+	filePath := filepath.Join(dir, name)
+	f, err := os.Create(filePath)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to create file: %s", filePath)
+	}
+	if err := reserveMetadataSpace(f); err != nil {
+		_ = os.Remove(filePath)
+		return nil, errors.Wrap(err, "failed to reserve metadata space in file")
+	}
+	if _, err := f.Seek(dataStartIndex, io.SeekStart); err != nil {
+		_ = os.Remove(filePath)
+		return nil, errors.Wrap(err, "failed to seek to data start in file")
+	}
+
+	return f, nil
+}
+
+// reserveMetadataSpace writes null bytes to the start of the file to reserve space for metadata and a newline as the last character.
+func reserveMetadataSpace(f *os.File) error {
+	buffer := make([]byte, headerEndIndex+1)
+	buffer[headerEndIndex] = headerEndByte
+	if _, err := f.WriteAt(buffer, 0); err != nil {
+		return errors.Wrapf(err, "failed to reserve metadata space in file")
+	}
+	return nil
+}
+
+// Write writes the given data to the file and appends a newline character.
+func (w *FileWriter) Write(ctx context.Context, data string) error {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	if w.stop {
+		return errors.New("cannot write to stopped writer")
+	}
+	_, err := w.f.WriteString(data + "\n")
+	if err != nil {
+		return errors.Wrapf(err, "failed to write to file")
+	}
+	w.metadata.Lines += w.increment
+	return nil
+}
+
+func (w *FileWriter) writeMetadata(ctx context.Context, metadata *Metadata) error {
+	if w.stop {
+		return errors.New("cannot write metadata to closed writer")
+	}
+	if _, err := w.f.WriteAt([]byte{metadataControlByte}, metadataControlByteIndex); err != nil {
+		return errors.Wrapf(err, "failed to write metadata control byte to file")
+	}
+	serialized := metadata.String()
+	length := byte(len(serialized))
```

Comment:
That means that the maximum length will be `255` (1 byte = uint8) - is that enough?

Reply:
Maybe 2 bytes is better, in theory, k8s names can be max 253 characters, so with 255 characters it can be breached